opal/mca/threads/base/tsd.c

```diff
@@ -0,0 +1,97 @@
+#include "opal/mca/threads/tsd.h"
+
+static void _tracked_destructor(void *arg) {
+    if (NULL == arg) {
+        return;
+    }
+
+    opal_tsd_list_item_t *tsd = (opal_tsd_list_item_t *)arg;
+
+    opal_tsd_tracked_key_t *key = tsd->tracked_key;
+  
+    opal_mutex_lock(&key->mutex);
+    opal_list_remove_item(&key->tsd_list, &tsd->super);
+    opal_mutex_unlock(&key->mutex);
+
+    if (NULL != key->user_destructor) {
+        key->user_destructor(tsd->data);
+    }
+
+    OBJ_RELEASE(tsd);
+}
+
+void opal_tsd_tracked_key_constructor(opal_tsd_tracked_key_t *key) {
+    OBJ_CONSTRUCT(&key->mutex, opal_mutex_t);
+    OBJ_CONSTRUCT(&key->tsd_list, opal_list_t);
+    key->user_destructor = NULL;
+    pthread_key_create(&key->key, &_tracked_destructor);
+}
+
+void opal_tsd_tracked_key_destructor(opal_tsd_tracked_key_t *key)
+{
+    opal_tsd_list_item_t *tsd;   
+
+    opal_mutex_lock(&key->mutex);
```

Comment:
```suggestion
```

Reply:
Its already assumed that no other thread can take the lock.
Otherwise it might get try to do that after mutex destruction